sys/net/gnrc/routing/simple_subnets/simple_subnets.c

```diff
@@ -0,0 +1,119 @@
+ */
+
+/**
+ * @defgroup    net_gnrc_simple_subnets Simple-Subnet auto-configuration
+ * @ingroup     net_gnrc
+ * @brief       Automatic configuration for cascading subnets
+ *
+ * About
+ * =====
+ *
+ * This module provides an automatic configuration for networks with a simple
+ * tree tolpology.
+ *
+ * If a sufficiently large IPv6 prefix is provided via Router Advertisements, a
+ * routing node with this module will automatically configure subnets from it for
+ * each downstream interface.
+ *
+ * There can only be a single routing node on each level of the network but an
+ * arbitrary number of leaf nodes.
+ *
+ * ![Example Topology](simple_subnets.svg)
+ *
+ * The downstream network(s) get the reduced prefix via Router Advertisements and
+ * the process repeats until the bits of prefix are exhausted. (The smallest
+ * must still have a /64 prefix.)
+ *
+ * The downstream router will send an otherwise empty router advertisement with only
+ * the Route Information Option set to the upstream network.
+ * The Route Information Option contains the prefix of the downstream network so that
+ * upstream hosts will no longer consider hosts in this subnet on-link but instead
+ * will use the downstream router to route to the new subnet.
+ *
+ * Usage
+ * =====
+ *
+ * Simply add the `gnrc_simple_subnets` module to the nodes that should act as routers
+ * in the cascading network.
+ * The upstream network will be automatically chosen as the one that first receives a
+ * router advertisement.
+ *
+ * @{
+ *
+ * @file
+ * @author  Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ */
+
+#include "net/gnrc/ipv6/nib.h"
+#include "net/gnrc/ndp.h"
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+static char addr_str[IPV6_ADDR_MAX_STR_LEN];
+
+void gnrc_ipv6_nib_rtr_adv_pio_cb(gnrc_netif_t *upstream, const ndp_opt_pi_t *pio)
+{
+    gnrc_netif_t *downstream = NULL;
+    gnrc_pktsnip_t *ext_opts = NULL;
+    const ipv6_addr_t *prefix = &pio->prefix;
+    uint32_t valid_ltime = byteorder_ntohl(pio->valid_ltime);
+    uint32_t pref_ltime = byteorder_ntohl(pio->pref_ltime);
+
+    /* create a subnet for each downstream interface */
+    unsigned subnets = gnrc_netif_numof() - 1;
+
+    const uint8_t prefix_len = pio->prefix_len;
+    uint8_t new_prefix_len;
+
+    if (subnets == 0) {
+        return;
+    }
+
+    new_prefix_len = prefix_len + 32 - __builtin_clz(subnets);
+
+    if (new_prefix_len > 64) {
+        DEBUG("simple_subnets: can't split /%u into %u subnets\n", prefix_len, subnets);
+        return;
+    }
+
+    while ((downstream = gnrc_netif_iter(downstream))) {
+        ipv6_addr_t new_prefix;
+
+        if (downstream == upstream) {
+            continue;
+        }
+
+        /* create subnet by adding interface index */
+        new_prefix.u64[0].u64 = byteorder_ntohll(prefix->u64[0]);
+        new_prefix.u64[0].u64 |= (uint64_t)subnets-- << (63 - prefix_len);
+        new_prefix.u64[0] = byteorder_htonll(new_prefix.u64[0].u64);
```

Comment:
Doesn't this immediately overwrite the value from the two lines before?

Reply:
Yea I first convert the prefix to host byte order so it's easier to work with.
Shifting the intext onto the network byte order address would cause trouble if it ends up between two bytes.

Reply:
Maybe add a comment to elevate code understanding.